lib/smart_proxy_remote_execution_ssh/runners/script_runner.rb

```diff
@@ -329,32 +330,36 @@ def run_sync(command, stdin = nil)
       stdout = ''
       stderr = ''
       exit_status = nil
-      started = false
 
-      channel = session.open_channel do |ch|
-        ch.on_data do |c, data|
-          stdout.concat(data)
-        end
-        ch.on_extended_data { |_, _, data| stderr.concat(data) }
-        ch.on_request('exit-status') { |_, data| exit_status = data.read_long }
-        # Send data to stdin if we have some
-        ch.send_data(stdin) unless stdin.nil?
-        # on signal: sending the signal value (such as 'TERM')
-        ch.on_request('exit-signal') do |_, data|
-          exit_status = data.read_string
-          ch.close
-          ch.wait
-        end
-        ch.exec command do |_, success|
-          raise 'could not execute command' unless success
-
-          started = true
-        end
+      options = []
+      options << "-o User=#{@ssh_user}"
+      options << "-o Port=#{ssh_options[:port]}" if ssh_options[:port]
+      options << "-o IdentityFile=#{ssh_options[:keys][0]}" if ssh_options[:keys]
+      options << "-o IdentitiesOnly=yes"
+      options << "-o StrictHostKeyChecking=accept-new"
+      options << "-o PreferredAuthentications=#{ssh_options[:auth_methods].join(',')}"
+      options << "-o UserKnownHostsFile=#{ssh_options[:user_known_hosts_file]}" if ssh_options[:user_known_hosts_file]
+      options << "-o NumberOfPasswordPrompts=#{ssh_options[:number_of_password_prompts]}"
+      options << "-o LogLevel=#{ssh_options[:verbose]}"
+
+      master, slave = PTY.open
+      read, write = IO.pipe
+      pid = spawn({"SSHPASS" => ssh_options[:password]}, "/usr/bin/sshpass", "-e", "/usr/bin/ssh", "#{@host}", *options, command, :in => read, :out => slave)
+      read.close
+      slave.close
+
+      write.puts(stdin) unless stdin.nil?
+      write.close
+
+      loop do
+        stdout += begin
+                    master.gets
+                  rescue Errno::EIO
```

Comment:
when does this happen? if it's "expected" could you add a comment when it happens and why? I assume we could see that if the target closes the connection unexpectedly. It would be great to test this and power off the target VM or unplug the cable etc. We may see other exceptions as well.

Reply:
It happens when reading from a file descriptor which is already closed. It won't be needed once https://github.com/theforeman/smart_proxy_remote_execution_ssh/pull/63#discussion_r681489452 is addressed.

The expected connection interruption will need to be handled elsewhere